### PR TITLE
Create fourbar subsytem

### PR DESCRIPTION
**Pull request by MrHodan**
Created the fourbar subsystem.

### CODE REVIEW

**Comment by MrHodan** [comment]
Gradle build broke, removing request and redoing

**Comment by MrHodan** [comment]
fixed gradle build

**Comment by Colepng** [comment]
Looks good to me

File: src/main/java/ca/warp7/frc2023/subsystems/Fourbar.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    public CANSparkMax motorSecondary =
+            SparkMaxUtil.createFollowerMotor(motorMain, FOURBAR_MOTOR_PORT_1, true); // need inversion, stops death
+
+    // the encoder (incremental)
+    public RelativeEncoder encoder = motorMain.getAlternateEncoder(
+            SparkMaxAlternateEncoder.Type.kQuadrature, 8192); // last value = count per rotation
+
+    // create and set a PID controller
+    public SparkMaxPIDController controllermotorMain =
+            SparkMaxUtil.createPIDController(motorMain, encoder, 0.1, 1e-4, 1, 0, 2000, 1500);
```
Comment (glowmouse) [inline src/main/java/ca/warp7/frc2023/subsystems/Fourbar.java]: I think this is

```
p = 0.1
i = 1e-4
d = 1
kMinVel = 0
kMaxVel = 2000
kMaxAcc = 1500
```

I think it'll be hard to tune the PID with an initial d value of 1.  I'd start with d = 0

File: src/main/java/ca/warp7/frc2023/subsystems/Fourbar.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+public class Fourbar extends SubsystemBase {
+    private static Fourbar instance;
+
+    public static Fourbar getInstance() {
+        if (instance == null) instance = new Fourbar();
+        return instance;
+    }
+
+    // creates a motorgroup to hold both motors, this is important as different commands could break the fourbar
+    public CANSparkMax motorMain = SparkMaxUtil.createMasterMotor(FOURBAR_MOTOR_PORT_0);
```
Comment (glowmouse) [inline src/main/java/ca/warp7/frc2023/subsystems/Fourbar.java]: Suggestion:  make these private if they're not going to be used outside the sub system.

File: src/main/java/ca/warp7/frc2023/subsystems/Fourbar.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+            SparkMaxUtil.createPIDController(motorMain, encoder, 0.1, 1e-4, 1, 0, 2000, 1500);
+
+    // gets the number of rotations from the encoder
+    public double getPosition() {
+        return encoder.getPosition();
+    }
+
+    // Sets the setPoint of the PID controller to whatever was passed as the input
+    public void setPosition(double length) {
+        controllermotorMain.setReference(length, CANSparkMax.ControlType.kPosition);
```
Comment (glowmouse) [inline src/main/java/ca/warp7/frc2023/subsystems/Fourbar.java:49]: I'm flagging something that I wasn't sure about...  I noticed that the controller type here, kPosition isn't he same as the type in the setReference at line 56 or so of SparkMaxUtil::createPIDController

```
        PIDController.setOutputRange(kMinOutput, kMaxOutput);
        PIDController.setReference(0, CANSparkMax.ControlType.kSmartMotion);
        PIDController.setSmartMotionMaxVelocity(kMaxVel, smartMotionSlot);
```

I looked at the class documentation for SparkMaxPIDController, and couldn't actually find a description of what these constants did.  The descriptions are blank: https://codedocs.revrobotics.com/java/com/revrobotics/cansparkmax.controltype.  :(

I found some example code:

https://github.com/REVrobotics/SPARK-MAX-Examples/blob/master/Java/Smart%20Motion%20Example/src/main/java/frc/robot/Robot.java

and that example seemed to use kSmartMotion to set the reference on a position PID.

File: src/main/java/ca/warp7/frc2023/subsystems/Fourbar.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    }
+}
+
+// please forgive our sins sir daniel
```
Comment (glowmouse) [inline src/main/java/ca/warp7/frc2023/subsystems/Fourbar.java:59]: Suggestion from something I saw in https://github.com/REVrobotics/SPARK-MAX-Examples/blob/master/Java/Smart%20Motion%20Example/src/main/java/frc/robot/Robot.java

You could add a periodic function and have it update the smart dashboard.  i.e.,

```
@Override
public void periodic() {
    SmartDashboard.putNumber("Fourbar Position", getPosition());
}
```

File: src/main/java/ca/warp7/frc2023/subsystems/Fourbar.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    // creates a motorgroup to hold both motors, this is important as different commands could break the fourbar
+    public CANSparkMax motorMain = SparkMaxUtil.createMasterMotor(FOURBAR_MOTOR_PORT_0);
+    public CANSparkMax motorSecondary =
+            SparkMaxUtil.createFollowerMotor(motorMain, FOURBAR_MOTOR_PORT_1, true); // need inversion, stops death
+
+    // the encoder (incremental)
+    public RelativeEncoder encoder = motorMain.getAlternateEncoder(
+            SparkMaxAlternateEncoder.Type.kQuadrature, 8192); // last value = count per rotation
```
Comment (glowmouse) [inline src/main/java/ca/warp7/frc2023/subsystems/Fourbar.java]: I'm flagging something that I wasn't sure about...

I read some of the documentation here: 
 https://docs.revrobotics.com/sparkmax/operating-modes/alternate-encoder-mode

and I think the alternate encoder means we're not using the motor encoder and, instead, hooking encoder hardware up to the the data port on the motor.  i.e.,

image

If that's the build, I was mostly curious about why the motor encoder wasn't used.

**Review by glowmouse — COMMENTED** [review]
All my comments are either "suggest only" feedback, or things I wasn't sure about when I read the code.

**Comment by MrHodan** [comment]
Added requested features, also fixed the gradle thing

File: build.gradle [hunk 6]
```diff
@@ -1,6 +1,6 @@
 plugins {
     id "java"
-    id "edu.wpi.first.GradleRIO" version "2023.3.2"
```
Comment (Dangosaurus) [inline build.gradle:3]: In the future you don't have to bump the GradleRIO version. I do that directly to the main repo so just rebase the change onto your fork.

File: src/main/java/ca/warp7/frc2023/lib/SparkMaxUtil.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
+package ca.warp7.frc2023.lib;
+
+import com.revrobotics.CANSparkMax;
+import com.revrobotics.CANSparkMaxLowLevel;
+import com.revrobotics.MotorFeedbackSensor;
+import com.revrobotics.SparkMaxPIDController;
+import edu.wpi.first.wpilibj.Encoder;
+
+public class SparkMaxUtil {
```
Comment (daniel-leejh) [inline src/main/java/ca/warp7/frc2023/lib/SparkMaxUtil.java]: Reiterating... Please avoid creating a lib file for something like this. You can just configure your motors in a method in your subsystem.

File: build.gradle [hunk 8]
```diff
@@ -1,6 +1,6 @@
...
+    id "edu.wpi.first.GradleRIO" version "2023.4.1"
```
Comment (daniel-leejh) [inline build.gradle]: In the future you don't have to bump the GradleRIO version. I do that directly to the main repo so just rebase the change onto your fork.

File: src/main/java/ca/warp7/frc2023/subsystems/Fourbar.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+import ca.warp7.frc2023.lib.SparkMaxUtil;
+import com.revrobotics.CANSparkMax;
+import com.revrobotics.RelativeEncoder;
+import com.revrobotics.SparkMaxPIDController;
+import edu.wpi.first.wpilibj.Encoder;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+
+public class Fourbar extends SubsystemBase {
```
Comment (daniel-leejh) [inline src/main/java/ca/warp7/frc2023/subsystems/Fourbar.java:13]: Please use javadoc comments in the future for any methods.

**Review by daniel-leejh — APPROVED** [review]
Merging into main for now...

@Katebin Please apply changes from the comments whenever you can. Thanks!

File: src/main/java/ca/warp7/frc2023/Constants.java [hunk 10]
```diff
@@ -1,3 +1,8 @@
 package ca.warp7.frc2023;
 
-public class Constants {}
+public class Constants {
+
+    // fourbar constants
+    public static final int FOURBAR_MOTOR_PORT_0 = 0;
```
Comment (daniel-leejh) [inline src/main/java/ca/warp7/frc2023/Constants.java]: Oops I missed this in the review... The naming convention that we use for comments is lowercase k then camel case of the rest of the name. E.g. kFourbarMotorID

**Comment by daniel-leejh** [comment]
@Katebin nvm not merging yet... There are build errors so might as well resolve the comments. Thanks!

**Pull request merged**